astroid/bases.py

```diff
@@ -119,9 +118,9 @@ def __getattr__(self, name):
             return self.__dict__[name]
         return getattr(self._proxied, name)
 
-    def infer(
-        self: _ProxyT, context: InferenceContext | None = None, **kwargs: Any
-    ) -> collections.abc.Generator[_ProxyT, None, None]:
+    def infer(  # type: ignore[return]
```

Comment:
`mypy` complains about wanting a `return None` statement whereas `pylint` complains about `useless-return`. Who is right here?

Reply:
I don't think you **need** an explicit return at the end of a generator, so I think mypy is wrong... But they say in their doc:

> You should give a statically typed function an explicit None return type even if it doesn’t return a value, as this lets mypy catch additional type errors:

So you must be explicit to help mypy it seems.

Reply:
I think that refers to:
```python
def func(x: str):
    print(x)
```

Which should have a `-> None` as well.

Reply:
Yeah I'll create a small reproducer in the coming days.

Reply:
Reported at https://github.com/python/mypy/issues/13019.

Reply:
@cdce8p See `mypy` issue. This is working as intended. Should we update `pylint` to allow `return None` for annotated functions?

Reply:
> But the real question is what should the default value be. mypy is used a lot so probably the one mypy chose to reduce friction.

I would agree. Both checks (mypy and pylint) are opinionated. The only thing for us is to make sure they don't collide if a users chooses to enable either one.

> Should we update `pylint` to allow `return None` for annotated functions?

Would it make sense to go one step further and disable `useless-return` entirely by default?

Reply:
> Would it make sense to go one step further and disable useless-return entirely by default?

Yeap, we don't need an option, my bad.

Reply:
We should really tweak the default values of `pylint`.

Reply:
>We should really tweak the default values of pylint.

Agreed. I'm thinking about applying 3512 but there's a lot of other issues to handle and I never have the time. This is probably one of the big pain point of pylint right now though. The issue that take us a lot of time can be very niche in very specific cases sometime but default values are really freaking important. (Also invalid-name for < 3 char value. I get bitten by it for new project myself, first thing I need to do is always to disable this warning, someone trying pylint always talk to me about it, it's not a good look).